crates/sui-core/src/authority/consensus_quarantine.rs

```diff
@@ -330,3 +334,647 @@ impl ConsensusCommitOutput {
+                default_per_commit_budget,
+            )
+        };
+        let shared_input_object_ids: BTreeSet<_> = transactions
+            .iter()
+            .filter_map(|tx| {
+                if let SequencedConsensusTransactionKind::External(ConsensusTransaction {
+                    kind: ConsensusTransactionKind::CertifiedTransaction(tx),
+                    ..
+                }) = &tx.0.transaction
+                {
+                    Some(tx.shared_input_objects().map(|obj| obj.id))
+                } else {
+                    None
+                }
+            })
+            .flatten()
+            .collect();
+
+        let num_ids = shared_input_object_ids.len();
+        let mut results = Vec::with_capacity(num_ids);
+        let mut fallback_keys = Vec::with_capacity(num_ids);
+        let mut fallback_indices = Vec::with_capacity(num_ids);
+
+        for (i, object_id) in shared_input_object_ids.iter().enumerate() {
+            if let Some(debt) = hash_table.get(object_id) {
+                results.push(Some(debt.into_v1()));
+            } else {
+                results.push(None);
+                fallback_keys.push(object_id);
+                fallback_indices.push(i);
+            }
+        }
+
+        let fallback_results = db_table.multi_get(fallback_keys)?;
+        assert_eq!(fallback_results.len(), fallback_indices.len());
+        for (i, result) in fallback_indices.into_iter().zip(fallback_results) {
+            results[i] = result.map(|debt| debt.into_v1());
+        }
+
+        Ok(results
+            .into_iter()
+            .zip(shared_input_object_ids)
+            .filter_map(|(debt, object_id)| debt.map(|debt| (debt, object_id)))
+            .map(move |((round, debt), object_id)| {
+                // Stored debts already account for the budget of the round in which
+                // they were accumulated. Application of budget from future rounds to
+                // the debt is handled here.
+                assert!(current_round > round);
+                let num_rounds = current_round - round - 1;
+                let debt = debt.saturating_sub(per_commit_budget * num_rounds);
+                (object_id, debt)
+            }))
+    }
+}
+
+// A wrapper around HashMap that uses refcounts to keep entries alive until
+// they are no longer needed.
+//
+// If there are N inserts for the same key, the key will not be removed until
+// there are N removes.
+//
+// It is intended to track the *latest* value for a given key, so duplicate
+// inserts are intended to overwrite any prior value.
+#[derive(Debug, Default)]
+struct RefCountedHashMap<K, V> {
+    map: HashMap<K, (usize, V)>,
+}
+
+impl<K, V> RefCountedHashMap<K, V>
+where
+    K: Clone + Eq + std::hash::Hash,
+{
+    pub fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+        }
+    }
+
+    pub fn insert(&mut self, key: K, value: V) {
+        let entry = self.map.entry(key);
+        match entry {
```

Comment:
would the more idiomatic pattern here be to use `and_modify` followed by `or_insert`?

Reply:
maybe but then you move value twice because the compiler can't see that only one of and_modify and or_insert will be called